### ISSUE TITLE
Cannot update popups via CmdlineChanged without a redraw

### ISSUE THREAD

**Issue by waivek** [vim/vim#8311]
**Describe the bug**

I am trying to add a popup component to my vim command line plugins. When the user types after pressing `:`, I want to parse their command and update popups accordingly.

Currently, updating a popup using the `CmdlineChanged` autocommand is not seamless. Changing the position and contents of a popup via `popup_move()` and `popup_settext()` do not work unless followed by an explicit `:redraw` command.

If the popup is also initially created via `CmdlineChanged`, it covers the input of the user. The fix for this is quite random. You need to start a timer using the `CmdlineChanged` command. I discovered this fix as I noticed the bug would come and go depending on my utilization of the `traces.vim` plugin, which makes heavy use of timers.

I have tried this on gvim and vim with a fresh vimrc and all plugins disabled. Would be grateful for any help.

**To Reproduce**

(This is a GIF below which replicates the different states of the bug.)

1. Download the attached `popup_bug.vim.txt` file.
2. Rename `popup_bug.vim.txt` to `popup_bug.vim`
3. Start gvim using the command `gvim -u .\popup_bug.vim -U .\popup_bug.vim --noplugin .\popup_bug.vim` 
4. The `popup_bug.vim` file contains a plugin that replicates what is being typed on the command line in a popup. This is a very stripped down version of my actual plugin, with no actual useful functionality other than to showcase my issue.
5. Type `echo "Attempt 1"`. The popup is drawn but we cannot see it.
6. Uncomment the two `redraw` commands in the file. Source the file via `:so %`.
7. Type `echo "Attempt 2"`. The popup is drawn and we can see it, but we can no longer see what we are typing.
8. Uncomment the line at the end of the file which starts a timer that does nothing on `CmdlineChanged`. Source the file via `:so %`.
9. Type `echo "Attempt 3"`. Now we can see both the command line and the popup being updated live.

**Expected behavior**

Ideally, I would like the effects of `popup_settext()` and `popup_move()` to be updated even with the `CmdlineChange` autocommand, without the requirement of the `:redraw` command. 

Manual `redraw` commands are expensive, and result in sluggish performance by introducing latency between 10ms-20ms *per keystroke/invocation*. This adds up and makes the typing experience unpleasant.

**Screenshots**

![runthrough v02](https://user-images.githubusercontent.com/5134623/120498242-b4e41500-c3dc-11eb-9f40-9aebeb72b33e.gif)

**Environment:**
 - Vim version 8.2
 - OS: Windows 10 20H2
 - Terminal: GUI

**Additional context**

This bug is not GUI exclusive. I have replicated it on the WSL CLI environment as well.

[popup_bug.vim.txt](https://github.com/vim/vim/files/6585070/popup_bug.vim.txt)

**Comment by brammool** [comment]
This is not going to be easy to solve.  When editing the command line, windows are not updated.  They may even be scrolled up, in which case redrawing them doesn't really work.  And popup windows are windows.  It might be possible to redraw the popup, but if it moved or changed size, what the draw in the screen space that it no longer covers?

I wonder what you use popup windows for while editing the command line.  Perhaps we should make a special kind of popup window for this.

**Comment by waivek** [comment]
I was making a plugin to traverse and filter the `v:oldfiles` variable, basically an MRU plugin. I had managed to make it using inbuilt vim features such as `:command` and the `command-complete` functions along with `wildcharm` for tabbing through the results. 

I wanted to further enhance the plugin by giving a preview of what would be selected _while_ the user was typing the command, _before_ activating the `wildmenu`. So I would parse what the user is typing and filter the oldfiles to get the first match of the pattern as it was being typed.

I've attached a gif of the plugin in action, along with the plugin file itself. 

Other ideas I had for popup windows while typing would be:
1. When the user types `:b <pattern>`, we could have a live preview of the file / the filename that would be selected using popups. 
2. When the user cycles through the `wildmenu`, different popups could be created to give additional information to the completion options if the `customlist` feature is being used. This would additionally require some sort of way to detect the user cycling through the `wildmenu`.
3. If we have case of a `vimgrep`, while the search pattern and file wildcard are being typed, we could have a popup to indicate how many files match the file wildcard. So the user would not have to actually execute the query to be aware of an incorrect wildcard.

Basically, the popup-while-typing-on-cmdline can be used to preemptively inform the user of the correctness of their command/query without executing it.

**The MRU Plugin**

![mru_example v01](https://user-images.githubusercontent.com/5134623/120533740-b0c8ef00-c3fe-11eb-84ee-afd486eca8c6.gif)

The plugin works like the `:b` command. 

1. If you simply enter `MRU` it opens `v:oldfiles[0]`, which is also the preview you get.
2. If you enter `MRU <pattern>` and press `ENTER`, you get the first `v:oldfiles` entry that matches the pattern. 
3. In either case, if you press `TAB`, you activate the `customlist` completion of the `:command` and cycle through the entries that match the pattern.

The plugin is below:

[mru.vim.txt](https://github.com/vim/vim/files/6586733/mru.vim.txt)
